Add GRASS 7.8.0

adamjstewart opened the pull request:
The latest release of GRASS finally supports Python 3.

adamjstewart commented:
@Sinan81 do you use GRASS? If so, can you test this for me?

Sinan81 commented:
I've never used it before though I might be playing around with it in the following weeks. I think @pat-s might be the right person to ping for this (hope he doesn't mind us pinging him here)

pat-s commented:
Happy to help. However, running into this on the first try

```sh
==> Building grass [AutotoolsPackage]
==> Error: LockTimeoutError: Timed out waiting for lock.

/opt/spack/lib/spack/spack/package.py:1376, in _stage_and_write_lock:
       1374    @contextlib.contextmanager
       1375    def _stage_and_write_lock(self):
  >>   1376        """Prefix lock nested in a stage."""
       1377        with self.stage:
       1378            with spack.store.db.prefix_write_lock(self.spec):
       1379                yield

```

adamjstewart commented:
Hmm, that's odd, and definitely not related to this PR. @tgamblin knows more about locks than I do.

adamjstewart commented:
@pat-s Did you figure out your weird locking issue? If not, you can use `spack -L` to disable locks in the meantime. This should be fine as long as you are the only person running this particular Spack instance and you don't run Spack simultaneously in multiple terminals.

pat-s commented:
Yes, it was gone lately. Idk what happened.

[Here](https://gist.github.com/pat-s/2751b0105094b98513794fd818808632) is the build log for me. Haven't yet taken a closer look, no time right now. Maybe it helps you.

adamjstewart commented:
@pat-s is this problem specific to GRASS 7.8.0 or Python 3? Can you try building 7.8.0 with Python 2 or building 7.6.1 if you have time?

That error message:
```
Fatal Python error: initfsencoding: unable to load the file system codec
ModuleNotFoundError: No module named 'encodings'
```
occurs when `PYTHONHOME` or `PYTHONPATH` are incorrectly set. Maybe the Makefile sets one of them?

Sinan81 commented:
When I try to install grass@7.8.0, I am getting the following error:
```
==> Installing grass
==> Searching for binary cache of grass
==> Warning: No Spack mirrors are currently configured
==> No binary for grass found: installing from source
==> Fetching https://grass.osgeo.org/grass78/source/grass-7.8.0.tar.gz
######################################################################## 100.0%
==> Staging archive: /cache/sbulut/spack-stage/grass-7.8.0-adhskgwhzlnwjhf3o7gzrldubqpgptut/grass-7.8.0.tar.gz
==> Created stage in /cache/sbulut/spack-stage/grass-7.8.0-adhskgwhzlnwjhf3o7gzrldubqpgptut
==> No patches needed for grass
==> Building grass [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/cache/sbulut/spack-stage/grass-7.8.0-adhskgwhzlnwjhf3o7gzrldubqpgptut/spack-src/configure' '--prefix=/mnt/local/spack/opt/spack/linux-centos7-skylake/gcc-7.4.0/grass-7.8.0-adhskgwhzlnwjhf3o7gzrldubqpgptut' '--without-odbc' '--without-nls' '--without-opendwg' '--with-x' '--with-gdal=/mnt/local/spack/opt/spack/linux-centos7-skylake/gcc-7.4.0/gdal-3.0.1-34rn5qjpbbga352db2z4sjl6k3opbscd/bin/gdal-config' '--with-proj-share=/mnt/local/spack/opt/spack/linux-centos7-skylake/gcc-7.4.0/proj-6.1.0-46z7xgenlwfcbb2wya3ohkhn3rj756oi/share/proj' '--with-cxx' '--with-tiff' '--with-png' '--without-postgres' '--without-mysql' '--with-sqlite' '--with-opengl' '--with-fftw' '--without-blas' '--without-lapack' '--with-cairo' '--with-freetype' '--without-readline' '--with-regex' '--without-pthread' '--without-openmp' '--without-opencl' '--without-bzlib' '--without-netcdf' '--without-geos'

1 error found in build log:
     93    checking for location of zlib library...
     94    checking for deflate in -lz... yes
     95    checking whether to use bzlib... no
     96    checking whether to use zstd... yes
     97    checking for location of zstd includes...
     98    checking for zstd.h... no
  >> 99    configure: error: *** Unable to locate zstd includes.

See build log for details:
  /cache/sbulut/spack-stage/grass-7.8.0-adhskgwhzlnwjhf3o7gzrldubqpgptut/spack-build-out.txt
```
This is on Centos7.  it seems zstd enabled by default and accordingly the dependency and the configuration options should be added.

adamjstewart commented:
Okay, I've fallen pretty far down the rabbit hole, so I added myself as a maintainer. I added new variants for things like `zstd` which was causing @Sinan81's problem, and new dependencies for things like `six` which was causing @pat-s's problem. I also switched most of the variants to be off by default, as I don't want to keep increasing the size of the DAG. I'm currently trying to build a minimal `grass~cxx~gdal ^proj@:5` but I'm getting the following errors:
```
Error: /Users/Adam/spack/lib/spack/env/clang/clang -E       -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -D__GLIBC_HAVE_LONG_LONG: In file included from /var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/tmpv2w62v5u.h:1:
Error: /Users/Adam/spack/lib/spack/env/clang/clang -E       -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -D__GLIBC_HAVE_LONG_LONG: In file included from /private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include/grass/gis.h:24:
Error: /Users/Adam/spack/lib/spack/env/clang/clang -E       -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -D__GLIBC_HAVE_LONG_LONG: In file included from /usr/include/stdio.h:64:
Error: /Users/Adam/spack/lib/spack/env/clang/clang -E       -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -D__GLIBC_HAVE_LONG_LONG: In file included from /usr/include/_stdio.h:68:
Error: /Users/Adam/spack/lib/spack/env/clang/clang -E       -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -D__GLIBC_HAVE_LONG_LONG: /usr/include/sys/cdefs.h:81:2: warning: "Unsupported compiler detected" [-W#warnings]
Error: /Users/Adam/spack/lib/spack/env/clang/clang -E       -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -D__GLIBC_HAVE_LONG_LONG: #warning "Unsupported compiler detected"
Error: /Users/Adam/spack/lib/spack/env/clang/clang -E       -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -D__GLIBC_HAVE_LONG_LONG:  ^
Error: /Users/Adam/spack/lib/spack/env/clang/clang -E       -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -I/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/grass-7.8.0-scvtnuaxtiq472v44lwiviui7hact6ig/spack-src/dist.x86_64-apple-darwin18.7.0/include -D__GLIBC_HAVE_LONG_LONG: 1 warning generated.
Status: Parsing /var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/tmpv2w62v5u.h
Error: /usr/include/sys/stdio.h:39: Syntax error at '__attribute__'
Error: /usr/include/sys/stdio.h:39: Syntax error at '='
Error: /usr/include/sys/stdio.h:46: Syntax error at '__attribute__'
Error: /usr/include/sys/stdio.h:46: Syntax error at '='
Error: /usr/include/sys/stdio.h:46: Syntax error at '='
Error: /usr/include/sys/stdio.h:46: Syntax error at '='
Error: /usr/include/sys/stdio.h:46: Syntax error at '='
Error: /usr/include/sys/stdio.h:47: Syntax error at '__attribute__'
Error: /usr/include/sys/stdio.h:47: Syntax error at '='
Error: /usr/include/sys/stdio.h:47: Syntax error at '='
Error: /usr/include/sys/stdio.h:47: Syntax error at '='
Error: /usr/include/sys/stdio.h:47: Syntax error at '='
Error: /usr/include/_stdio.h:137: Syntax error at '_close'
Error: /usr/include/_stdio.h:138: Syntax error at '_read'
Error: /usr/include/_stdio.h:139: Syntax error at '_seek'
Error: /usr/include/_stdio.h:140: Syntax error at '_write'
Error: /usr/include/_stdio.h:140: Syntax error at ','
Error: /usr/include/_stdio.h:157: Syntax error at '}'
Error: /usr/include/stdio.h:142: Syntax error at '*'
Error: /usr/include/stdio.h:143: Syntax error at '*'
Error: /usr/include/stdio.h:144: Syntax error at '*'
Error: /usr/include/stdio.h:145: Syntax error at '*'
Error: /usr/include/stdio.h:146: Syntax error at '*'
Error: /usr/include/stdio.h:147: Syntax error at '*'
Error: /usr/include/stdio.h:148: Syntax error at '*'
Error: /usr/include/stdio.h:149: Syntax error at 'FILE'
Error: /usr/include/stdio.h:153: Syntax error at '__filename'
Error: /usr/include/stdio.h:153: Syntax error at '__mode'
Error: /usr/include/stdio.h:155: Syntax error at '*'
Error: /usr/include/stdio.h:155: Syntax error at '...'
Error: /usr/include/stdio.h:155: Syntax error at 'i2'
Error: /usr/include/stdio.h:156: Syntax error at 'FILE'
Error: /usr/include/stdio.h:157: Syntax error at 'FILE'
Error: /usr/include/stdio.h:158: Syntax error at '__ptr'
Error: /usr/include/stdio.h:158: Syntax error at 'size_t'
Error: /usr/include/stdio.h:160: Syntax error at 'FILE'
Error: /usr/include/stdio.h:161: Syntax error at '*'
Error: /usr/include/stdio.h:161: Syntax error at '...'
Error: /usr/include/stdio.h:161: Syntax error at 'i2'
Error: /usr/include/stdio.h:162: Syntax error at '*'
Error: /usr/include/stdio.h:163: Syntax error at '*'
Error: /usr/include/stdio.h:163: Syntax error at ')'
Error: /usr/include/stdio.h:164: Syntax error at '*'
Error: /usr/include/stdio.h:165: Syntax error at '__ptr'
Error: /usr/include/stdio.h:165: Syntax error at 'size_t'
Error: /usr/include/stdio.h:166: Syntax error at '*'
Error: /usr/include/stdio.h:170: Syntax error at '__attribute__'
Error: /usr/include/stdio.h:170: Syntax error at 'i1'
Error: /usr/include/stdio.h:171: Syntax error at 'FILE'
Error: /usr/include/stdio.h:176: Syntax error at '*'
Error: /usr/include/stdio.h:177: Syntax error at '__attribute__'
Error: /usr/include/stdio.h:177: Syntax error at 'i1'
Error: /usr/include/stdio.h:178: Syntax error at '*'
Error: /usr/include/stdio.h:178: Syntax error at ')'
Error: /usr/include/stdio.h:179: Syntax error at '*'
Error: /usr/include/stdio.h:179: Syntax error at 'int'
Error: /usr/include/stdio.h:180: Syntax error at '__attribute__'
Error: /usr/include/stdio.h:180: Syntax error at 'i2'
Error: /usr/include/stdio.h:180: Syntax error at '='
Error: /usr/include/stdio.h:181: Syntax error at '__attribute__'Status: Preprocessing /var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/tmp8zjgry2d.h
...
```
I've never seen this error before. Maybe it's because I'm building on macOS? Can anyone try on Linux?

adamjstewart commented on var/spack/repos/builtin/packages/grass/package.py:
```diff
@@ -194,13 +240,18 @@ def configure_args(self):
-        return None
+        if self.spec.satisfies('@:7.6'):
+            makefile = FileFilter('include/Make/Platform.make')
+            makefile.filter(r'^ICONVLIB\s*=\s*', 'ICONVLIB = -liconv')
```
This was causing linking errors like "no library found: -liconv-liconv", so I assume it was fixed in 7.8.0.

Sinan81 replied:
The error you report above convinces me that 'iconv' issue is fixed in the new version. if you disable it for 7.8 and don't get any missing library 'iconv' error, then it should be fine.

Sinan81 replied:
Actually, I just tried compiling 7.8 it seems in my case ICONVLIB is blank as opposed to being equla to '-iconv' as was also the case for version 7.6. So, for centos it seems to be necessary to do the filtering of Platform.make.

Sinan81 commented:
After enabling Platform.make filtering for version 7.8, the next error I am getting is:
```
Fatal Python error: Py_Initialize: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings'
```
This is for python@2.7.16. Will also try building with python3 shortly.

Sinan81 commented:
I am not getting the above mentioned import error in Python 3. However, the next error has to do with compilation of 'display/d.where' component of grass
```
sbulut@ws-067 /mnt/local/cache/sbulut/spack-stage/grass-7.8.0-k6tct6xworxarcdwo5ehrqevseoz37yq/spack-src/display/d.where 
$ make
/mnt/local/spack/lib/spack/env/gcc/gcc  -g -O2   -I/mnt/local/cache/sbulut/spack-stage/grass-7.8.0-k6tct6xworxarcdwo5ehrqevseoz37yq/spack-src/dist.x86_64-pc-linux-gnu/include -I/mnt/local/cache/sbulut/spack-stage/grass-7.8.0-k6tct6xworxarcdwo5ehrqevseoz37yq/spack-src/dist.x86_64-pc-linux-gnu/include    -DPACKAGE=\""grassmods"\"    -I/mnt/local/cache/sbulut/spack-stage/grass-7.8.0-k6tct6xworxarcdwo5ehrqevseoz37yq/spack-src/dist.x86_64-pc-linux-gnu/include -I/mnt/local/cache/sbulut/spack-stage/grass-7.8.0-k6tct6xworxarcdwo5ehrqevseoz37yq/spack-src/dist.x86_64-pc-linux-gnu/include -DRELDIR=\"display/d.where\" -o OBJ.x86_64-pc-linux-gnu/main.o -c main.c
main.c: In function 'main':
main.c:88:2: warning: implicit declaration of function 'exit' [-Wimplicit-function-declaration]
  exit(EXIT_FAILURE);
  ^~~~
main.c:88:2: warning: incompatible implicit declaration of built-in function 'exit'
main.c:88:2: note: include '<stdlib.h>' or provide a declaration of 'exit'
main.c:88:7: error: 'EXIT_FAILURE' undeclared (first use in this function); did you mean 'UNIT_FILE'?
  exit(EXIT_FAILURE);
       ^~~~~~~~~~~~
       UNIT_FILE
main.c:88:7: note: each undeclared identifier is reported only once for each function it appears in
main.c:175:5: warning: incompatible implicit declaration of built-in function 'exit'
     exit(EXIT_SUCCESS);
     ^~~~
main.c:175:5: note: include '<stdlib.h>' or provide a declaration of 'exit'
main.c:175:10: error: 'EXIT_SUCCESS' undeclared (first use in this function)
     exit(EXIT_SUCCESS);
```

Sinan81 commented:
As suggested, the error above goes away after adding `#include <stdlib.h>` to display/d.where/main.c and rest of the build works without errors.

adamjstewart commented:
Hmm, I wonder if GRASS 7.8.0 _only_ supports Python 3?

pat-s commented:
> @pat-s is this problem specific to GRASS 7.8.0 or Python 3? Can you try building 7.8.0 with Python 2 or building 7.6.1 if you have time?

Is this still valid or already outdated?

adamjstewart commented:
@pat-s probably outdated, but you're welcome to try again with my latest fixes.

pat-s commented:
Here is the most recent [build log](https://transfer.sh/b4BPa/spack-build-out.txt).

You guys probably know this, but for completion:

`spack install grass ^proj@5.2.0 ^libgeotiff@1.4.3 ^gdal@2.4.2` 

- proj@5 because grass requires proj@5
- libgeotiff@4 because libgeotiff@5 requires proj@6 (which clashes with proj@5 from grass)
- gdal@2.4.2 because gdal@3 requires proj@6 (which clashes with proj@5 from grass)

pat-s commented:
@adamjstewart Builds for me on spack v0.13.0 using `variants: +geos ~opengl ~freetype`

adamjstewart commented:
So @pat-s is able to build this, but last I checked, I’m not and @Sinan81 isn’t either. I’ll try again, but I want things to be a little more stable before I merge.

pat-s commented:
Furthermore, it seems that adjustments are required for all versions < 7.8.0 as the formula now defaults to py-numpy@1.17 using python3. However, grass < 7.8.0 cannot deal with this and needs py-numpy <= 1.16

```py
depends_on('python@2.7:2.8,3.5:', type=('build', 'run'), when='@1.16:')
depends_on('python@3.5:', type=('build', 'run'), when='@1.17:')
```

Also I was unable to build grass < 7.8.0 at all.

adamjstewart commented:
Yeah, that's a bug in the concretizer. @tgamblin is working on fixing that right now. I don't want to hardcode too many hacks in packages in the meantime.

adamjstewart commented:
Tried again and got the same errors. I noticed a "Unsupported compiler detected" warning, so I think the problem is with Apple Clang.

pat-s commented:
Could be. I was building on centOS7.

adamjstewart commented:
Hmm, my error message looks a lot like https://trac.osgeo.org/grass/ticket/3231, which was supposedly fixed. I wonder if it happened again?

adamjstewart commented:
Well, I'm seeing "Unsupported compiler detected" when I try GCC, so that's a red herring.
